pkg/core/writer/splunk/splunk.go

```diff
@@ -0,0 +1,240 @@
+	Index      string            `json:"index,omitempty"`      // optional name of the Splunk index to store the event in; not required if the token has a default index set in Splunk
+	Event      string            `json:"event"`                // type of event: this is a metric.
+	Fields     map[string]string `json:"fields"`               // metric data
+}
+
+type splunkEvent struct {
+	Time       int64     `json:"time"`                 // epoch time
+	Host       string    `json:"host"`                 // hostname
+	Source     string    `json:"source,omitempty"`     // optional description of the source of the event; typically the app's name
+	SourceType string    `json:"sourcetype,omitempty"` // optional name of a Splunk parsing configuration; this is usually inferred by Splunk
+	Index      string    `json:"index,omitempty"`      // optional name of the Splunk index to store the event in; not required if the token has a default index set in Splunk
+	Event      eventdata `json:"event"`                // event data
+}
+
+func toString(obj interface{}) string {
+	if stringer, ok := obj.(fmt.Stringer); ok {
+		return stringer.String()
+	}
+	return fmt.Sprintf("%v", obj)
+}
+
+func computeTime(timestamp time.Time) int64 {
+	if timestamp.IsZero() {
+		return time.Now().UnixNano() / time.Millisecond.Nanoseconds()
+	}
+	return timestamp.UnixNano() / time.Millisecond.Nanoseconds()
+}
+
+// LogDataPoint logs a data point as a Splunk metric event
+func (h *Writer) LogDataPoint(d *datapoint.Datapoint) {
+	fields := make(map[string]string)
+
+	for key, v := range d.Meta {
+		if v != nil {
+			fields[toString(key)] = toString(v)
+		}
+	}
+	for key, v := range d.Dimensions {
+		fields[toString(key)] = toString(v)
+	}
+	fields["metric_type"] = d.MetricType.String()
+	fields[fmt.Sprintf("metric_name:%s", d.Metric)] = d.Value.String()
+
+	newEvent := splunkMetric{
+		Time:       computeTime(d.Timestamp),
+		Host:       h.hostname,
+		Source:     h.source,
+		SourceType: h.sourceType,
+		Index:      h.index,
+		Event:      "metric",
+		Fields:     fields,
+	}
+	h.events <- newEvent
+}
+
+// LogEvent logs an event as a Splunk metric event
+func (h *Writer) LogEvent(e *event.Event) {
+	props := make(map[string]string)
+	for key, v := range e.Properties {
+		if v != nil {
+			props[key] = toString(v)
+		}
+	}
+
+	meta := make(map[string]string)
+	for key, v := range e.Meta {
+		if v != nil {
+			meta[toString(key)] = toString(v)
+		}
+	}
+	newEvent := splunkEvent{
+		Time:       computeTime(e.Timestamp),
+		Host:       h.hostname,
+		Source:     h.source,
+		SourceType: h.sourceType,
+		Index:      h.index,
+		Event:      eventdata{Properties: props, Dimensions: e.Dimensions, Meta: meta, EventType: e.EventType, Category: e.Category},
+	}
+	h.events <- newEvent
+}
+
+func (h *Writer) logEvents(events []interface{}) {
+	buf := new(bytes.Buffer)
+	for _, e := range events {
+		b, err := json.Marshal(e)
```

Comment:
I think it might be more efficient to use https://golang.org/pkg/encoding/json/#Encoder and pass in the buffer so that it doesn't have to allocate the separate byte array on heap.

Reply:
OK, changed.